# PRACTICE Exam 1: Part 2

Your acknowledgement of the honor code in Sakai applies to this notebook.

This is an individual assessment. This document and your answers are for your eyes only.

This exam is open book, open note, open Internet, and closed to people. This means you cannot receive help on this exam from anyone, including (but not limited to) communicating with a person while taking the exam, such as asking someone through the Internet (like stackoverflow) to receive help.

You may **not use** `for` or `while` loops for this part unless you can guarantee that the loop will iterate less than 25 times, including if we swapped out the data for another data file with 10x more rows. The use of such loops will immediately mark your solution as "Not Yet" because one large purpose of this exam is to assess your skills at taking advantage of the data science libraries.

The grading for this exam will be very lenient in the reasonableness of how you use the data since we do not expect you to have a great understanding of the data set. Moreover, the values in this data set are not necessarily reflective of the real world.

In [1]:
import numpy as np
import pandas as pd
import math

### Column Definitions
This section involves the `heart_attacks` DataFrame. Its columns are explained below.

1. `age` - Age of person
1. `sex` - 1 = male; 0 = female
1. `chest pain` - Type of chest pain
    1. Value 0: typical angina
    1. Value 1: atypical angina
    1. Value 2: non-anginal pain
    1. Value 3: asymptomatic
1. `blood pressure` - Resting blood pressure (in mm Hg)
1. `cholestoral` - Cholestoral in mg/dl fetched via BMI sensor
1. `blood sugar` - Fasting blood sugar > 120 mg/dl (1 = true; 0 = false)
1. `resting ecg` - Resting electrocardiographic results
    1. Value 0: normal
    1. Value 1: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV)
    1. Value 2: showing probable or definite left ventricular hypertrophy by Estes' criteria
1. `max heart rate` - Maximum heart rate achieved
1. `angina` - Exercise induced angina (1 = yes; 0 = no)
1. `old peak` - ST depression induced by exercise relative to rest
1. `slope` - The slope of the peak exercise ST segment
1. `vessel num` - number of major vessels (0-3) colored by flourosopy
1. `thall` - 3 = normal; 6 = fixed defect; 7 = reversable defect
1. `HA chance` - 0 = less chance of heart attack; 1 = more chance of heart attack

Note: The values in this data set are not necessarily reflective of the real world.

In [2]:
# Run this code to load the data
heart_attacks = pd.read_csv('heart.csv')
heart_attacks.head()

age  sex  chest pain  blood pressure  cholestoral  blood sugar  \
0   63    1           3             145          233            1   
1   37    1           2             130          250            0   
2   41    0           1             130          204            0   
3   56    1           1             120          236            0   
4   57    0           0             120          354            0   

   resting ecg  max heart rate  angina  old peak  slope  vessel num  thall  \
0            0             150       0       2.3      0           0      1   
1            1             187       0       3.5      0           0      2   
2            0             172       0       1.4      2           0      2   
3            1             178       0       0.8      2           0      2   
4            1             163       1       0.6      2           0      2   

   HA chance  
0          1  
1          1  
2          1  
3          1  
4          1

In [3]:
# Feel free to use this cell and add more to explore the data set
# BEGIN SOLUTION
print(heart_attacks['chest pain'].value_counts())

heart_attacks.describe(include='all')
# END SOLUTION

0    143
2     87
1     50
3     23
Name: chest pain, dtype: int64


age         sex  chest pain  blood pressure  cholestoral  \
count  303.000000  303.000000  303.000000      303.000000   303.000000   
mean    54.366337    0.683168    0.966997      131.623762   246.264026   
std      9.082101    0.466011    1.032052       17.538143    51.830751   
min     29.000000    0.000000    0.000000       94.000000   126.000000   
25%     47.500000    0.000000    0.000000      120.000000   211.000000   
50%     55.000000    1.000000    1.000000      130.000000   240.000000   
75%     61.000000    1.000000    2.000000      140.000000   274.500000   
max     77.000000    1.000000    3.000000      200.000000   564.000000   

       blood sugar  resting ecg  max heart rate      angina    old peak  \
count   303.000000   303.000000      303.000000  303.000000  303.000000   
mean      0.148515     0.528053      149.646865    0.326733    1.039604   
std       0.356198     0.525860       22.905161    0.469794    1.161075   
min       0.000000     0.000000       71.000000    0.000000    0.000000   
25%       0.000000     0.000000      133.500000    0.000000    0.000000   
50%       0.000000     1.000000      153.000000    0.000000    0.800000   
75%       0.000000     1.000000      166.000000    1.000000    1.600000   
max       1.000000     2.000000      202.000000    1.000000    6.200000   

            slope  vessel num       thall   HA chance  
count  303.000000  303.000000  303.000000  303.000000  
mean     1.399340    0.729373    2.313531    0.544554  
std      0.616226    1.022606    0.612277    0.498835  
min      0.000000    0.000000    0.000000    0.000000  
25%      1.000000    0.000000    2.000000    0.000000  
50%      1.000000    0.000000    2.000000    1.000000  
75%      2.000000    1.000000    3.000000    1.000000  
max      2.000000    4.000000    3.000000    1.000000

## Question 1 (20 points)

Create a mask that uses at least two columns and selects for at most 75% and at least 25% of the original data set. This mask must be of booleans, one for each row in the dataframe. Make that mask a new column. Use that mask to filter the data and from the resulting data calculate a single descriptive statistic for one of the columns. Finally, provide a written explanation of what the code does.

To help with grading, put the new column name in `new_col_name`, the fraction of values kept in `frac_data_kept`, and place your descriptive statistic in `statistic`. Make sure to leave and run the last few lines of the cell below for grading.

## Answer 1: Code

In [4]:
""" # BEGIN PROMPT
new_col_name = ...

frac_data_kept = ...

statistic = ...
""" # END PROMPT

# BEGIN SOLUTION NO PROMPT
# Exemplary Answer:
new_col_name = 'higher risk'
heart_attacks[new_col_name] = (heart_attacks['blood pressure'] > 120) & (heart_attacks['age'] > 55)
frac_data_kept = np.sum(heart_attacks[new_col_name])/len(heart_attacks)
statistic = np.mean(heart_attacks[heart_attacks[new_col_name]]['cholestoral'])


# Satisfactory Answer example 1:
# This code is not clean and easy to read because while the mask column is created it is never used
# instead the code that created the mask is copy and pasted multiple times
new_col_name = 'higher risk'
heart_attacks[new_col_name] = (heart_attacks['blood pressure'] > 120) & (heart_attacks['age'] > 55)
frac_data_kept = np.sum((heart_attacks['blood pressure'] > 120) & (heart_attacks['age'] > 55))/len(heart_attacks)
statistic = np.mean(heart_attacks[(heart_attacks['blood pressure'] > 120) & (heart_attacks['age'] > 55)]['cholestoral'])


# Satisfactory Answer example 2:
## Filters out only 1.65% data, while Exemplary requires 25%~75%
conditions = [(heart_attacks['age'] <= 35) & (heart_attacks['cholestoral'] >= 190)]
heart_attacks['High Cholestoral In Young Adults'] = np.select(conditions, [True])
heart_attacks['High Cholestoral In Young Adults'] = heart_attacks['High Cholestoral In Young Adults'].astype(bool)


total = len(heart_attacks)
portion = heart_attacks['High Cholestoral In Young Adults'].sum()

new_df = heart_attacks[heart_attacks['High Cholestoral In Young Adults'] == True]
sum_chol = new_df['cholestoral'].sum()
avg_chol = sum_chol / portion

new_col_name = 'High Cholestoral In Young Adults'

frac_data_kept = portion/total

statistic = avg_chol


# Not Yet Answer: 
## frac_data_kept is not a float
heart_attacks["MaleU60"] = np.where(heart_attacks['age'] < 60, np.where(heart_attacks['sex'] == 1, True, False),False)

frac_data_kept = heart_attacks[heart_attacks["MaleU60"]]

statistic = heart_attacks.loc[heart_attacks['MaleU60'] == True, 'blood pressure'].mean()

new_col_name = "MaleU60"
# END SOLUTION

# Leave these here to help with grading.
print('New column name:', new_col_name)
print('Fraction of data kepted:', frac_data_kept)
print('Descriptive statistic of new column:', statistic)

New column name: MaleU60
Fraction of data kepted:      age  sex  chest pain  blood pressure  cholestoral  blood sugar  \
1     37    1           2             130          250            0   
3     56    1           1             120          236            0   
5     57    1           0             140          192            0   
7     44    1           1             120          263            0   
8     52    1           2             172          199            1   
..   ...  ...         ...             ...          ...          ...   
291   58    1           0             114          318            0   
294   44    1           0             120          169            0   
297   59    1           0             164          176            1   
299   45    1           3             110          264            0   
301   57    1           0             130          131            0   

     resting ecg  max heart rate  angina  old peak  slope  vessel num  thall  \
1              1 

## Answer 1: Written portion

<!-- BEGIN QUESTION -->



_Type your answer here, replacing this text._

**Solution**

### Exemplary Answer

I created a new column called "higher risk" that is true for those that have a blood pressure higher than 120 and are over 55. 37.6% of those in the data set fit this criteria and their mean cholestoral is 252.

#### Rubric
 1. Creates a new column in the dataframe, the name is in the variable `new_col_name`
 2. The new column is a Series of booleans
 3. The mask is created using at least 2 columns
 4. The mask is 25-75% of the data set, and the fraction is in `frac_data_kept`
 5. Uses the mask to select rows/cells of a column
 6. There is a descriptive statistic of a different column in `statistic`
 7. The explanation matches the code
 1. The code is clean and easy to read

### Satisfactory Answer

I created a column called "High Cholestoral In Young Adults" that has a boolean value of True if the person is 35 and under while also having a cholestoral level that is a equal to or above 190 (which is considered to be entering the range of high cholestoral). A very small portion of the data set fits this criteria **(only ~1.65%)** and of these individuals, the average cholestorol level is 217.2.

#### Rubric
 - Needs #1, #3
 - #2 - can be series of 0's and 1's instead
 - #4 - can be out of range, but needs to be a float
 - #5 - can not have this and did it some other way (likely copy and pasting code)
 - #6 - there is a descriptive statistic but it is not of a different column
 - #7 - can be a little off
 - #8 - it is not clean or easy to read, examples: (1) rather than using the new column they copy and pasted the code to recreate the mask, (2) uses `== True` instead of just the mask

### Not Yet Answer

I created a new column called MaleU60 to check if the patient was a male under the age of 60. 51 percent of the patients fell under this category. Of the patients under 60 and male the average blood pressure was 129.53548387096774.

#### Rubric

 - Does not meet satisfactory, such as something else is in `new_col_name` rather than being of type `str` and has the new column's name, or #4 is not a float

#### Unassessable Rubric

 - Work that is missing, does not demonstrate meaningful effort, or does not provide enough evidence to determine a level of mastery.

<!-- END QUESTION -->

## Question 2 (20 points)

Create events A and B such that you can derive the empirical probabilities $P(A)$, $P(B)$, $P(A|B)$, and $P(B|A)$
using the data set. In the code cell below, write the code to directly compute all four empirical probabilities from the dataset. You should calculate every value here independently by manipulating the dataframe (for example, filter and select as appropriate) and then dividing two integer values.

Then, compute $P(A|B)$ and $P(B|A)$ again using Bayes' Theorem. When computing $P(A|B)$, you can use the empirical values of $P(A)$, $P(B)$, and $P(B|A)$ above, and vice versa. Confirm the values of $P(A|B)$ and $P(B|A)$ when calculated via Bayes' Theorem match the version you get from the dataframe.

Use the variables in the skeleton code below and leave the print statements for grading. In the markdown cell below, describe the two events.

## Answer 2: Code

In [5]:
""" # BEGIN PROMPT
...

prob_a = ...
prob_b = ...
prob_a_given_b_data = ...
prob_b_given_a_data = ...

prob_a_given_b_bayes = ...
prob_b_given_a_bayes = ...
""" # END PROMPT

# BEGIN SOLUTION NO PROMPT
# Exemplary Answer
nAll = len(heart_attacks)

nAttack = sum(heart_attacks['HA chance'] == 1)
hasAttack = heart_attacks[heart_attacks['HA chance'] == 1]

nFemale = sum(heart_attacks['sex'] == 0)
isFemale = heart_attacks[heart_attacks['sex'] == 0]

prob_a = nFemale/nAll
prob_b = nAttack/nAll
prob_a_given_b_data = sum(hasAttack['sex'] == 0)/nAttack
prob_b_given_a_data = sum(isFemale['HA chance'] == 1)/nFemale

prob_a_given_b_bayes = (prob_b_given_a_data*prob_a)/prob_b
prob_b_given_a_bayes = (prob_a_given_b_data*prob_b)/prob_a


# Satisfactory Answer
## From the code, we see that "high cholesterol" is defined as "cholesterol > 200", but the writeup is vague.
high_cholesterol = heart_attacks['cholestoral'] > 200
male = heart_attacks['sex'] == 1
## COPY-PASTING THE MASKS rather than using the above variables
## A clean code should re-use the variables: high_cholesterol_and_male = high_cholesterol & male
high_cholesterol_and_male = (heart_attacks['cholestoral'] > 200) & (heart_attacks['sex'] == 1)
prob_high_cholesterol_and_male = high_cholesterol_and_male.sum() / len(heart_attacks)

prob_a = high_cholesterol.sum() / len(heart_attacks)
prob_b = male.sum() / len(heart_attacks)

prob_a_given_b_data = prob_high_cholesterol_and_male / prob_b
prob_b_given_a_data = prob_high_cholesterol_and_male / prob_a

prob_a_given_b_bayes = (prob_b_given_a_data * prob_a) / prob_b
prob_b_given_a_bayes = (prob_a_given_b_data * prob_b) / prob_a


# Not Yet Answer
## Error in code
prob_a = heart_attacks[heart_attacks["blood pressure"] > 140].shape[0]/heart_attacks.shape[0]
prob_b = heart_attacks[heart_attacks["age"] > 60].shape[0]/heart_attacks.shape[0]

bp_above_140 = heart_attacks["blood pressure"] > 140
older_60 = heart_attacks["age"] > 60
prob_a_given_b_data = older_60[bp_above_140].shape[0]/heart_attacks[heart_attacks["age"] > 60].shape[0]
prob_b_given_a_data = bp_above_140[older_60].shape[0]/bp_above_140.shape[0]

prob_a_given_b_bayes = ...
prob_b_given_a_bayes = ...
# END SOLUTION

# Leave these here to help with grading.
print('prob_a', prob_a)
print('prob_b', prob_b)
print('prob_a_given_b_data', prob_a_given_b_data)
print('prob_b_given_a_data', prob_b_given_a_data)
print('prob_a_given_b_bayes', prob_a_given_b_bayes)
print('prob_b_given_a_bayes', prob_b_given_a_bayes)
print('Match? a_given_b:', math.isclose(prob_a_given_b_data, prob_a_given_b_bayes, abs_tol=0.00000001))
print('Match? b_given_a:', math.isclose(prob_b_given_a_data, prob_b_given_a_bayes, abs_tol=0.00000001))

prob_a 0.2145214521452145
prob_b 0.2607260726072607
prob_a_given_b_data 0.8227848101265823
prob_b_given_a_data 0.2607260726072607
prob_a_given_b_bayes Ellipsis
prob_b_given_a_bayes Ellipsis


TypeError: must be real number, not ellipsis

## Answer 2: Written portion

<!-- BEGIN QUESTION -->



_Type your answer here, replacing this text._

**solution**

### Exemplary Answer

$A$ = Has high chance of heart attack, $B$ = Is female

#### Rubric
 1. Clear defines events $A$ and $B$
 2. Correctly calculates `prob_a` and `prob_b`
 3. Correctly calculates `prob_a_given_b_data` and `prob_b_given_a_data` using only the data frame and not relying on probability formulas
 4. Correctly calculates `prob_a_given_b_bayes` and `prob_b_given_a_bayes` using Bayes formula
 5. $P(A|B)$ and $P(B|A)$ match (ignore floating point issues)
 1. The code is clean and easy to read

### Satisfactory Answer

A is having high cholesterol (**What is high? From code, it's defined as `cholesterol > 200`, which is "can be vague if clear from code"**) and B is male.

#### Rubric
 - Must have: #2, #3, #4, #5
 - #1 - can be vague if clear from code
 - #6 - it is not clean or easy to read, example: copy and pasting the mask instead of calculating once and then using it in a variable
 
### Not Yet Answer

A is blood pressure above 140 and B is age above 60. (**There are errors in the code section**)

#### Rubric
 - Work that does not meet some requirements and/or displays developing or incomplete mastery of at least some learning goals and material. Such as not doing #3 where it calculates with the data frame by just following the probability formula rather than counting N for when A and B are true and the N for when just A is true.

### Unassessable Rubric
 - Work that is missing, does not demonstrate meaningful effort, or does not provide enough evidence to determine a level of mastery.

<!-- END QUESTION -->

## Question 3 (20 points)

This data set contains both men and women. Filter the data set by some criteria such that you have 75% or less of the original data set. Report the proportion of women. Then assuming n is large enough that the the central limit theorem holds calculate how many observations would be needed to conclude that the probability of observing the empirical proportion is at most 5% (i.e., P(observe the empirical proportion) <= 0.05) when assuming the true proportion is 50%.

In the code cell below and using the provided variables (you may make more as you need):
1. Filter the data
1. Calculate the proportion of women
1. Calculate the number of needed observations

In the markdown cell below:
1. Explain how you filtered the data
1. Report whether your current sample has enough observations


## Question 3 (20 points)

Filter the data set by some criteria such that you have 75% or less of the original data set. Decide on a second criteria, then calculate the proportion of the filtered data where the second criteria is true. Then decide on what the true mean/proportion could be (50% or something else). Finally, calculate how many observations would be needed to conclude that the probability of observing the empirical proportion is at most 5% (i.e., P(observe the empirical proportion) <= 0.05) when assuming the true proportion. You can assume n is large enough that the central limit theorem holds.

In the code cell below do the following steps:
1. Filter the data
1. Calculate the empirical proportion
1. Calculate the number of needed observations

Fill in the following variables to help with grading (you may make more as you need):
1. `n_filtered` - The number of data points left after your code filters the data
1. `frac_filtered` - The fraction of data points that are left after filtering the data
1. `true_prop` - The true proportion you are using for your calculations
1. `prop_of_filtered` - The proportion of your filtered data where the second criteria is true
1. `variance` - The variance you are using for your calculations
1. `delta` - The delta you are using for your calculations
1. `n_needed` - The number of observations you need

In the markdown cell below state:
1. How you filtered the data
1. What your second proportion was
1. Describe what you used as your true mean and your rationale for your choice
1. Whether your current sample has enough observations

## Answer 3: Code

In [6]:
""" # BEGIN PROMPT
...

n_filtered = ...
frac_filtered = ...
true_prop = ...
prop_of_filtered = ...
variance = ...
delta = ...
n_needed = ...
""" # END PROMPT

# BEGIN SOLUTION NO PROMPT
# Exemplary Answer
high_ha = heart_attacks[heart_attacks['HA chance']==1]
n_filtered = len(high_ha)
frac_filtered = n_filtered/heart_attacks.shape[0]
prop_of_filtered = len(high_ha[high_ha['sex'] == 0])/len(high_ha)
true_prop = 0.5
variance = 0.25
delta = prop_of_filtered - 0.5
n_needed = (4*variance)/(delta**2)


# Satisfactory Answer
## n_filtered is a DataFrame
angina_mask = heart_attacks["angina"]==1
n_filtered = heart_attacks[angina_mask]
totalSize = n_filtered.shape[0]
num_women =(n_filtered[n_filtered["sex"]==0]).shape[0]
frac_filtered = totalSize/heart_attacks.shape[0]
prop_of_filtered = num_women/totalSize
true_prop = 0.5
variance = 0.25
delta = 0.5-prop_of_filtered
n_needed = (4*variance) / (delta*delta)


# Not Yet Answer
## Incorrectly Calculated Variance
angina = heart_attacks[heart_attacks['angina'] == 1]
f = angina[angina['sex']==0]
n_filtered = len(angina)
frac_filtered = n_filtered/len(heart_attacks)
prop_of_filtered = len(f)/n_filtered
true_prop = 0.5
variance = angina.var()['sex']
delta = 0.5 - prop_of_filtered
n_needed = (4 * variance) / (delta**2)

# END SOLUTION

# Leave these here to help with grading.
print('Size of filtered sample:', n_filtered)
print('Proportion of original data:', frac_filtered)
print('Proportion of filtered data:', prop_of_filtered)
print('True proportion:', true_prop)
print('Variance:', variance)
print('Delta:', delta)
print('Needed N:', n_needed)

Size of filtered sample: 99
Proportion of original data: 0.32673267326732675
Proportion of filtered data: 0.2222222222222222
True proportion: 0.5
Variance: 0.17460317460317473
Delta: 0.2777777777777778
Needed N: 9.051428571428577


## Answer 3: Written portion

<!-- BEGIN QUESTION -->



_Type your answer here, replacing this text._

**Solution**

### Exemplary Answer

I first filtered by those with a high chance of a heart attack. 43.6% of the filtered data are women. I define the true mean as 50% because in broad generalities half of the world's population is men and the other half women and there are only two genders in this data set. Assuming that n is large enough that the central limit theorem holds,  we would need 247 observations to conclude that the probability of observing 43.6% is at most 5%. However, we only have 165 samples, so we do not have enough.

#### Rubric
 1. Explanation matches the code
 2. Filter criteria results in <= 75% of the data set and `n_filtered` is properly defined.
 3. Correctly calculates all of the other variables
 4. Rationale for true mean choice is reasonable
 4. Explanation includes observation of whether there are enough samples
 5. Prints proportion of original data and it is stated
 6. The code is clean and easy to read
 
 
### Satisfactory Answer
I filtered the data by whether they had an exercise induced angina and then looked at the proportion of women vs men. I defined the true proportion as 50% since the data set is using a binary gender. I calculated that we would need 13 samples to have enough observations to conclude that the probability of observing the empirical proportion is at most 5%. Since the data I filtered for has 99 samples we do have enough observations. (**Code part assigns `n_filtered` to a DataFrame**)

#### Rubric
 - Must have: #3
 - #1 - can be missing some stuff if code clearly has all of the pieces and none are incorrect
 - #2 - can be >75% as long as the filter does remove some of them, can be a print of the data frame
 - #4 - rationale can be weak
 - #5 - can be missing
 - #6 - can be incorrect (like `n_filtered` is a dataframe) and/or missing from written portion
 - #7 - it is not clean or easy to read, example: copy and pasting the mask instead of calculating once and then using it in a variable


### Not Yet Answer

From the given data, we observe that 22.22% of those who get angina when they exercise are women. If we are assuming that n is large enough that the central limit theorem holds and the true proportion is 50%, we would need 10 observations to conclude that the P(of observing 22.22%) <= 5%. The current example we have does have enough samples as there are 99 samples and we only need 10. Looking at the data this makes sense to me because the delta is large indicating that the value of n itself will be smaller. (**Variance incorrectly calculated and no mention of why chose 50% as true proportion**)

#### Rubric

 - Work that does not meet some requirements and/or displays developing or incomplete mastery of at least some learning goals and material, such as incorrectly calculating some of the variables for rubric #3.
 
#### Unassessable Rubric
 - Work that is missing, does not demonstrate meaningful effort, or does not provide enough evidence to determine a level of mastery.

<!-- END QUESTION -->

